pkg/controllers/monitoring/monitoring-stack/conditions_test.go

```diff
@@ -176,10 +194,68 @@ func TestUpdateConditions(t *testing.T) {
 					Reason:             PrometheusNotReconciled,
 				}},
 		},
+		{
+			name: "Prometheus observed generation is different to the Prometheus generation",
+			originalMSConditions: []v1alpha1.Condition{
+				{
+					Type:               v1alpha1.AvailableCondition,
+					Status:             v1alpha1.ConditionTrue,
+					ObservedGeneration: 2,
+					Reason:             AvailableReason,
+					Message:            AvailableMessage,
+					LastTransitionTime: transitionTime,
+				},
+				{
+					Type:               v1alpha1.ReconciledCondition,
+					Status:             v1alpha1.ConditionTrue,
+					ObservedGeneration: 2,
+					Reason:             ReconciledReason,
+					Message:            SuccessfullyReconciledMessage,
+					LastTransitionTime: transitionTime,
+				},
+			},
+			generation: 2,
+			recError:   nil,
+			prometheus: monv1.Prometheus{
+				ObjectMeta: metav1.ObjectMeta{
+					Generation: 3,
+				},
+				Status: monv1.PrometheusStatus{
+					Conditions: []monv1.PrometheusCondition{
+						{
+							Type:               monv1.PrometheusAvailable,
+							Status:             monv1.PrometheusConditionFalse,
+							ObservedGeneration: 2,
+						},
+						{
+							Type:               monv1.PrometheusReconciled,
+							Status:             monv1.PrometheusConditionFalse,
+							ObservedGeneration: 2,
+						},
+					}}},
+			expectedResults: []v1alpha1.Condition{
+				{
+					Type:               v1alpha1.AvailableCondition,
+					Status:             v1alpha1.ConditionTrue,
+					ObservedGeneration: 2,
+					Reason:             AvailableReason,
+					Message:            AvailableMessage,
+					LastTransitionTime: transitionTime,
+				},
+				{
+					Type:               v1alpha1.ReconciledCondition,
+					Status:             v1alpha1.ConditionTrue,
+					ObservedGeneration: 2,
+					Reason:             ReconciledReason,
+					Message:            SuccessfullyReconciledMessage,
+					LastTransitionTime: transitionTime,
+				}},
+			sameTransitionTimes: true,
+		},
 	}
 
 	for _, test := range tt {
-		res := updateConditions(test.originalMSConditions, test.prometheusConditions, test.generation, test.recError)
+		res := updateConditions(test.originalMSConditions, test.prometheus, test.generation, test.recError)
```

Comment:
Would these unit tests suffice? I wonder if we should add an e2e test that tests this scenario in real world. 
I believe having an e2e validates this specific requirement is met and can potentially catch an error in PO and/or ObO should things change later.

Reply:
An e2e test would be great indeed. It'd mean to scale the prometheus operator to zero, modify the Prometheus resource and check that the status doesn't change.

Reply:
Hmm does it mean we need to know the namespace where the OBO is installed (to be able to scale down the P-O)? Can we find it deterministically? I can imagine we could use hardcoded value for kind, but it's not very nice....other option is to list all the deployments and try to find by name (again not very nice IMO). Or am I missing something?

Reply:
> It'd mean to scale the prometheus operator to zero ... 

That may not be an option since OLM can possibly reset it albeit it doesn't seem to at the moment but we shouldn't rely on this behaviour. 

Would it possible to somehow set the `ms.Spec.PrometheusConfig` to be something invalid that results in PO not updating the ObservedGeneration ?

Reply:
Hmm ok I can try....but I guess it will be a bit tricky. I think the time when there will be the difference between the generations will be very very short.

Reply:
Yes I think Simon is right that we would need to scale down the P-O. I was experimenting manually and when you update Prometheus config then the P-O updates the status conditions (observedGeneration) almost immediately. When I scaled down the P-O then I was able to reproduce such situation.

>Would it possible to somehow set the ms.Spec.PrometheusConfig to be something invalid that results in PO not updating the ObservedGeneration ?

I guess P-O should always update the status or not? 

Anyway I would suggested to merge this PR and then maybe work on the e2e test.

Reply:
I'd be ok with merging this without the e2e test. Iiuc this test would mostly test correct behavior of PO anyway?

Reply:
I think it's not just PO behaviour. We want to be sure OBO works as expected, but otherwise yes....let's work on the e2e test in different PR and don't block this one.